test/unit/sessions.test.js

```diff
@@ -181,9 +160,148 @@ describe('Sessions - unit/core', function () {
     });
+
+    describe('new ClientSession()', () => {
+      it('should throw errors with invalid parameters', function () {
+        expect(() => {
+          new ClientSession();
+        }).to.throw(/ClientSession requires a topology/);
+
+        expect(() => {
+          new ClientSession({});
+        }).to.throw(/ClientSession requires a ServerSessionPool/);
+
+        expect(() => {
+          new ClientSession({}, {});
+        }).to.throw(/ClientSession requires a ServerSessionPool/);
+      });
+
+      it('should throw an error if snapshot and causalConsistency options are both set to true', function () {
+        const client = new Topology('localhost:27017', {});
+        sessionPool = client.s.sessionPool;
+        expect(
+          () => new ClientSession(client, sessionPool, { causalConsistency: true, snapshot: true })
+        ).to.throw('Properties "causalConsistency" and "snapshot" are mutually exclusive');
+      });
+
+      it('should default to `null` for `clusterTime`', function () {
+        const client = new Topology('localhost:27017', {});
+        sessionPool = client.s.sessionPool;
+        session = new ClientSession(client, sessionPool);
+        expect(session.clusterTime).to.not.exist;
+      });
+
+      it('should set the internal clusterTime to `initialClusterTime` if provided', function () {
+        const clusterTime = genClusterTime(Date.now());
+        const client = new Topology('localhost:27017');
+        sessionPool = client.s.sessionPool;
+        session = new ClientSession(client, sessionPool, { initialClusterTime: clusterTime });
+        expect(session.clusterTime).to.eql(clusterTime);
+      });
+
+      it('should acquire a serverSession in the constructor if the session is explicit', () => {
+        const session = new ClientSession(topology, serverSessionPool, { explicit: true });
+        const serverSessionSymbol = getSymbolFrom(session, 'serverSession');
+        expect(session).to.have.property(serverSessionSymbol).that.is.an.instanceOf(ServerSession);
+      });
+
+      it('should leave serverSession null if the session is implicit', () => {
+        // implicit via false (this should not be allowed...)
+        let session = new ClientSession(topology, serverSessionPool, { explicit: false });
+        const serverSessionSymbol = getSymbolFrom(session, 'serverSession');
+        expect(session).to.have.property(serverSessionSymbol, null);
+        // implicit via omission
+        session = new ClientSession(topology, serverSessionPool, {});
+        expect(session).to.have.property(serverSessionSymbol, null);
+      });
+
+      it('should start the txnNumberIncrement at zero', () => {
+        const session = new ClientSession(topology, serverSessionPool);
+        const txnNumberIncrementSymbol = getSymbolFrom(session, 'txnNumberIncrement');
+        expect(session).to.have.property(txnNumberIncrementSymbol, 0);
+      });
+    });
+
+    describe('get serverSession()', () => {
```

Comment:
Now that we a clearly defined behavior for the getter, I think we should split up the cases to cover the following scenarios:
- session symbol is non-null, the session is explicit, regardless of session ended or not
- session symbol is non-null, the session is implicit, regardless of session ended or not
- session symbol is null, the session is explicit, regardless of session ended or not
- session symbol is null, the session is implicit, session not ended
- session symbol is null, the session is implicit, session is ended